apps/src/dropletUtils.js

```diff
@@ -404,7 +417,15 @@ exports.generateAceApiCompleter = function (functionFilter, dropletConfig) {
         callback(null, []);
         return;
       }
-      callback(null, apis);
+      var line = session.getLine(pos.row);
+      var identifier = acUtil.retrievePrecedingIdentifier(line, pos.column);
+      if (pos.column > (identifier.length + 1) &&
+          line[pos.column - identifier.length - 1] === '.') {
```

Comment:
I'm not sure I follow this, so I'm going to try and walk through it in text.  I had to look up the doc for [retrievePrecedingIdentifier](https://github.com/ajaxorg/ace/blob/5c79952b29e5f2ccb5f8ecc779e80b07e1406951/lib/ace/autocomplete/util.js#L50).  Correct me wherever I go wrong here.

```javascript
var line = session.getLine(pos.row);
```
**line** {string} is the complete text of the editor line containing the cursor.

```javascript
var identifier = acUtil.retrievePrecedingIdentifier(line, pos.column);
```
**identifier** {string} should be the text of the complete symbol preceding the cursor position - not including any periods.  If the character directly before the cursor position is not a valid symbol character, this will be empty string.
* I think `pos.column` is the cursor position within the line, where 0 is before the first character.
* `acUtil` means 'autocomplete utility' right? Not mistyped 'aceUtil'

Examples:

> If `line` is `var x = obj` and the cursor is at the end of the line (column 11) then identifier should be `"obj"`.

> If `line` is `var x = obj.` and the cursor is at the end of the line (column 12) then identifier should be `""`.

I'll break up the conditional on the next line into two parts

```javascript
pos.column > (identifier.length + 1);
```
I'm not sure what to name this condition if I were to extract it.  It might be checking for whether the cursor is one character beyond the end of the most recent symbol, but since `pos.column` is relative to the whole line and `identifier` is already just the most recent identifier, I think this gives a false positive in some situations.

Examples:

> If `line` is `obj.` and the cursor is before the "j" (column 2), then `identifier` is "ob"
  `2 > ("ob".length + 1); // false` - this case seems correct

> If `line` is `obj.` and the cursor is at the end of the line (column 4), then `identifier` is ""
  `4 > ("".length + 1); // true` - also correct, but maybe by accident?

> If `line` is `var x = obj.` and the cursor is before the "j" (column 10), then `identifier` is "ob"
  `10 > ("ob".length + 1); // true` - this seems wrong

Finally:
```javascript
line[pos.column - identifier.length - 1] === '.'
```

Again, column is relative to the line, but the identifier may not go back to the beginning of the line, so I'm not sure what this checks.

Examples:

> If `line` is `obj.` and the cursor is before the "j" (column 2), then `identifier` is "ob"
  `line[2 - "ob".length - 1] // line[-1]` - gives `undefined` which makes the expression false, which works accidentally.

> If `line` is `obj.` and the cursor is at the end of the line (column 4), then `identifier` is ""
  `line[4 - "".length - 1] // line[3]` - gives "." which makes the expression true, which works accidentally.

> If `line` is `var x = obj.` and the cursor is before the "j" (column 10), then `identifier` is "ob"
  `line[10 - "ob".length - 1] // line[7]` - gives " " which is false, which works accidentally.

Again, I'm not 100% sure I followed this correctly, so please correct anything above.  I just... I think this might work, but I really don't understand _how_.